apps/emqx_rule_engine/src/emqx_rule_engine.erl

```diff
@@ -19,627 +19,107 @@
-                  emqx_rule_registry:get_resources()).
-
-refresh_resource(Type) when is_atom(Type) ->
-    lists:foreach(fun refresh_resource/1,
-                  emqx_rule_registry:get_resources_by_type(Type));
-
-refresh_resource(#resource{id = ResId}) ->
-    emqx_rule_monitor:ensure_resource_retrier(ResId, ?T_RETRY).
-
--spec(refresh_rules() -> ok).
-refresh_rules() ->
-    lists:foreach(fun
-        (#rule{enabled = true} = Rule) ->
-            try refresh_rule(Rule)
-            catch _:_ ->
-                emqx_rule_registry:add_rule(Rule#rule{enabled = false, state = refresh_failed_at_bootup})
-            end;
-        (_) -> ok
-    end, emqx_rule_registry:get_rules()).
-
-refresh_rule(#rule{id = RuleId, for = Topics, actions = Actions}) ->
-    ok = emqx_rule_metrics:create_rule_metrics(RuleId),
-    lists:foreach(fun emqx_rule_events:load/1, Topics),
-    refresh_actions(Actions).
-
--spec(refresh_resource_status() -> ok).
-refresh_resource_status() ->
-    lists:foreach(
-        fun(#resource{id = ResId, type = ResType}) ->
-            case emqx_rule_registry:find_resource_type(ResType) of
-                {ok, #resource_type{on_status = {Mod, OnStatus}}} ->
-                    _ = fetch_resource_status(Mod, OnStatus, ResId);
-                _ -> ok
-            end
-        end, emqx_rule_registry:get_resources()).
-
 %%------------------------------------------------------------------------------
 %% Internal Functions
 %%------------------------------------------------------------------------------
-prepare_actions(Actions, NeedInit) ->
-    [prepare_action(Action, NeedInit) || Action <- Actions].
 
-prepare_action(#{name := Name, args := Args0} = Action, NeedInit) ->
-    case emqx_rule_registry:find_action(Name) of
-        {ok, #action{module = Mod, on_create = Create, params_spec = ParamSpec}} ->
-            Args = emqx_rule_validator:validate_params(Args0, ParamSpec),
-            ActionInstId = maps:get(id, Action, action_instance_id(Name)),
-            case NeedInit of
-                true ->
-                    _ =  emqx_plugin_libs_rule:cluster_call(?MODULE, init_action, [Mod, Create, ActionInstId,
-                            with_resource_params(Args)]),
-                    ok;
-                false -> ok
-            end,
-            #action_instance{
-                id = ActionInstId, name = Name, args = Args,
-                fallbacks = prepare_actions(maps:get(fallbacks, Action, []), NeedInit)
-            };
-        not_found ->
-            throw({action_not_found, Name})
-    end.
-
-with_resource_params(Args = #{<<"$resource">> := ResId}) ->
-    case emqx_rule_registry:find_resource_params(ResId) of
-        {ok, #resource_params{params = Params}} ->
-            maps:merge(Args, Params);
-        not_found ->
-            throw({resource_not_initialized, ResId})
-    end;
-with_resource_params(Args) -> Args.
+%% The pattern {'ok', Select} can never match the type {'error',{_,[{_,_,_,_}]}}.
```

Comment:
we need this.
https://github.com/emqx/rulesql/commit/7e08693a23bd946491a1dcf3988f050917a492e3

which is tagged `0.1.3`